src/libraries/Microsoft.PowerFx.Core/Texl/Builtins/Collect.cs

```diff
@@ -0,0 +1,472 @@
+                        if (!dataSourceType.TryGetType(name, out DType dsNameType))
+                        {
+                            dataSourceType.ReportNonExistingName(FieldNameKind.Display, errors, name, args[i], DocumentErrorSeverity.Warning);
+                        }
+                    }
+                }
+            }
+
+            base.CheckSemantics(binding, args, argTypes, errors);
+            base.ValidateArgumentIsMutable(binding, args[0], errors);            
+
+            MutationUtils.CheckSemantics(binding, this, args, argTypes, errors);
+
+            if (binding.Features.PowerFxV1CompatibilityRules)
+            {
+                MutationUtils.CheckForReadOnlyFields(argTypes[0], args.Skip(1).ToArray(), argTypes.Skip(1).ToArray(), errors);
+            }
+        }
+
+        // This method returns true if there are special suggestions for a particular parameter of the function.
+        public override bool HasSuggestionsForParam(int argumentIndex)
+        {
+            Contracts.Assert(argumentIndex >= 0);
+
+            return argumentIndex == 0;
+        }
+
+        public override bool TryGetDataSourceNodes(PowerFx.Syntax.CallNode callNode, TexlBinding binding, out IList<FirstNameNode> dsNodes)
+        {
+            Contracts.AssertValue(callNode);
+            Contracts.AssertValue(binding);
+
+            dsNodes = new List<FirstNameNode>();
+            if (callNode.Args.Count != 2)
+            {
+                return false;
+            }
+
+            var args = Contracts.VerifyValue(callNode.Args.Children);
+            var arg1 = Contracts.VerifyValue(args[1]);
+
+            // Only the second arg can contribute to the output for the purpose of delegation
+            return ArgValidators.DataSourceArgNodeValidator.TryGetValidValue(arg1, binding, out dsNodes);
+        }
+
+        public override IEnumerable<Identifier> GetIdentifierOfModifiedValue(TexlNode[] args, out TexlNode identifierNode)
+        {
+            Contracts.AssertValue(args);
+
+            identifierNode = null;
+            if (args.Length == 0)
+            {
+                return null;
+            }
+
+            var firstNameNode = args[0]?.AsFirstName();
+            identifierNode = firstNameNode;
+
+            if (firstNameNode == null)
+            {
+                return null;
+            }
+
+            var identifiers = new List<Identifier>
+            {
+                firstNameNode.Ident
+            };
+            return identifiers;
+        }
+
+        public override bool IsAsyncInvocation(PowerFx.Syntax.CallNode callNode, TexlBinding binding)
+        {
+            Contracts.AssertValue(callNode);
+            Contracts.AssertValue(binding);
+
+            return Arg0RequiresAsync(callNode, binding);
+        }
+
+        public static DType GetCollectedTypeForGivenArgType(Features features, DType argType, TexlNode arg, ref Dictionary<TexlNode, DType> nodeToCoercedTypeMap)
```

Comment:
>GetCollectedTypeForGivenArgType [](http://example.com/codeflow?start=28&length=31)

This is only used in CollectScalarFunction; does it need to be in this base class?